### PR TITLE
UNDERTOW-1469 UndertowOutputStream.resetBuffer properly resets the buffer

### PR DESCRIPTION
**Pull request by cakofony**

### CODE REVIEW

**Comment by undertow-pull-request** [comment]
**Windows** [Build 2627](https://ci.wildfly.org/viewLog.html?buildId=134585&buildTypeId=UNDERTOW_MasterWindows) outcome was **FAILURE** using a merge of c69c7ff6677fb096b05008ebbf0c7363ed6b388f
Summary: Tests failed: 2 (2 new), passed: 3365, ignored: 709 Build time: 00:43:42
### Failed tests
```
io.undertow.server.handlers.ReceiverTestCase.testAsyncReceiveWholeBytesFailed[proxy][http2]: java.lang.AssertionError
	at io.undertow.server.handlers.ReceiverTestCase.testAsyncReceiveWholeBytesFailed(ReceiverTestCase.java:194)
	at io.undertow.testutils.DefaultServer.runChild(DefaultServer.java:524)
	at io.undertow.testutils.DefaultServer.runChild(DefaultServer.java:104)
	at io.undertow.testutils.DefaultServer.run(DefaultServer.java:297)
------- Stderr: -------
10:38:56,567 DEBUG (XNIO-1 I/O-6) [io.undertow.server.handlers.proxy] <ProxyHandler.java:549> Sending request ClientRequest{path='/fullbytes', method=POST, protocol=HTTP/1.1} to target localhost/127.0.0.1:8888 for exchange HttpServerExchange{ POST /fullbytes request {Content-Length=[130000100], Host=[localhost]} response {}}
10:38:56,567 DEBUG (XNIO-1 I/O-6) [io.undertow.server.handlers.proxy] <ProxyHandler.java:566> Sent request ClientRequest{path='/fullbytes', method=POST, protocol=HTTP/1.1} to target 127.0.0.1 for exchange HttpServerExchange{ POST /fullbytes request {Content-Length=[130000100], Host=[localhost]} response {}}
10:38:56,605 ERROR (XNIO-1 I/O-6) [io.undertow.proxy] <ProxyHandler.java:662> UT005028: Proxy request to /fullbytes failed: java.nio.channels.ClosedChannelException
	at io.undertow.client.http2.Http2ClientConnection$1.handleEvent(Http2ClientConnection.java:107)
	at io.undertow.client.http2.Http2ClientConnection$1.handleEvent(Http2ClientConnection.java:99)
	at org.xnio.ChannelListeners.invokeChannelListener(ChannelListeners.java:92)
	at io.undertow.server.protocol.framed.AbstractFramedChannel$FrameCloseListener.handleEvent(AbstractFramedChannel.java:1061)
	at io.undertow.server.protocol.framed.AbstractFramedChannel$FrameCloseListener.handleEvent(AbstractFramedChannel.java:975)
	at org.xnio.ChannelListeners.invokeChannelListener(ChannelListeners.java:92)
	at io.undertow.protocols.ssl.SslConduit$SslReadReadyHandler.terminated(SslConduit.java:1214)
	at org.xnio.nio.NioSocketConduit.readTerminated(NioSocketConduit.java:360)
	at org.xnio.nio.NioSocketStreamConnection.notifyReadClosed(NioSocketStreamConnection.java:148)
	at org.xnio.Connection.close(Connection.java:139)
	at org.xnio.IoUtils.safeClose(IoUtils.java:134)
	at io.undertow.protocols.ssl.SslConduit.closed(SslConduit.java:1046)
	at io.undertow.protocols.ssl.SslConduit.close(SslConduit.java:1117)
	at io.undertow.protocols.ssl.UndertowSslConnection.closeAction(UndertowSslConnection.java:149)
	at org.xnio.Connection.close(Connection.java:132)
	at org.xnio.IoUtils.safeClose(IoUtils.java:134)
	at io.undertow.server.protocol.framed.AbstractFramedChannel.close(AbstractFramedChannel.java:816)
	at org.xnio.IoUtils.safeClose(IoUtils.java:134)
	at io.undertow.protocols.http2.Http2Channel.sendGoAway(Http2Channel.java:832)
	at io.undertow.protocols.http2.Http2Channel.sendGoAway(Http2Channel.java:809)
	at io.undertow.client.http2.Http2ClientConnection.close(Http2ClientConnection.java:303)
	at org.xnio.IoUtils.safeClose(IoUtils.java:134)
	at io.undertow.server.handlers.proxy.ProxyHandler$IoExceptionHandler.handleException(ProxyHandler.java:790)
	at org.xnio.ChannelListeners.invokeChannelExceptionHandler(ChannelListeners.java:126)
	at io.undertow.util.Transfer$TransferListener.handleEvent(Transfer.java:218)
	at org.xnio.ChannelListeners.invokeChannelListener(ChannelListeners.java:92)
	at io.undertow.server.protocol.framed.AbstractFramedStreamSinkChannel$1.run(AbstractFramedStreamSinkChannel.java:214)
	at io.undertow.server.protocol.framed.AbstractFramedChannel$FrameReadListener.handleEvent(AbstractFramedChannel.java:935)
	at io.undertow.server.protocol.framed.AbstractFramedChannel$FrameReadListener.handleEvent(AbstractFramedChannel.java:929)
	at org.xnio.ChannelListeners.invokeChannelListener(ChannelListeners.java:92)
	at org.xnio.conduits.ReadReadyHandler$ChannelListenerHandler.readReady(ReadReadyHandler.java:66)
	at io.undertow.protocols.ssl.SslConduit$SslReadReadyHandler.readReady(SslConduit.java:1175)
	at io.undertow.protocols.ssl.SslConduit$1.run(SslConduit.java:168)
	at org.xnio.nio.WorkerThread.safeRun(WorkerThread.java:582)
	at org.xnio.nio.WorkerThread.run(WorkerThread.java:466)

10:38:56,605 DEBUG (XNIO-1 I/O-6) [io.undertow.request.error-response] <DelegatingBasicLogger.java:424> Setting error code 503 for exchange HttpServerExchange{ POST /fullbytes request {Content-Length=[130000100], Host=[localhost]} response {}}: java.lang.RuntimeException
	at io.undertow.server.HttpServerExchange.setStatusCode(HttpServerExchange.java:1410)
	at io.undertow.server.handlers.proxy.ProxyHandler.handleFailure(ProxyHandler.java:668)
	at io.undertow.server.handlers.proxy.ProxyHandler$ResponseCallback.failed(ProxyHandler.java:727)
	at io.undertow.client.http2.Http2ClientExchange.failed(Http2ClientExchange.java:125)
	at io.undertow.client.http2.Http2ClientConnection$1.handleEvent(Http2ClientConnection.java:107)


io.undertow.websockets.jsr.test.JsrWebSocketServer07Test.testCloseFrame: java.lang.IllegalStateException: ChannelPipeline does not contain a HttpRequestEncoder or HttpClientCodec
	at io.netty.handler.codec.http.websocketx.WebSocketClientHandshaker$1.operationComplete(WebSocketClientHandshaker.java:182)
	at io.netty.handler.codec.http.websocketx.WebSocketClientHandshaker$1.operationComplete(WebSocketClientHandshaker.java:172)
	at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:507)
	at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:481)
	at io.netty.util.concurrent.DefaultPromise.access$000(DefaultPromise.java:34)
	at io.netty.util.concurrent.DefaultPromise$1.run(DefaultPromise.java:431)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:403)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:445)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
	at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:144)
	at java.lang.Thread.run(Thread.java:748)
------- Stderr: -------
11:08:45,404 DEBUG (XNIO-1 I/O-6) [io.undertow.request.io] <AbstractFramedChannel.java:876> Marking writes broken on channel WebSocket13Channel peer localhost/127.0.0.1:7777 local 0.0.0.0/0.0.0.0:52160[ No Receiver [] -- [] -- []]: java.nio.channels.ClosedChannelException
	at io.undertow.server.protocol.framed.AbstractFramedChannel.markWritesBroken(AbstractFramedChannel.java:876)
	at io.undertow.server.protocol.framed.AbstractFramedStreamSinkChannel.channelForciblyClosed(AbstractFramedStreamSinkChannel.java:562)
	at io.undertow.server.protocol.framed.AbstractFramedStreamSinkChannel.close(AbstractFramedStreamSinkChannel.java:539)
	at org.xnio.IoUtils.safeClose(IoUtils.java:134)
	at io.undertow.server.protocol.framed.AbstractFramedChannel.queueFrame(AbstractFramedChannel.java:706)
	at io.undertow.server.protocol.framed.AbstractFramedStreamSinkChannel.queueFinalFrame(AbstractFramedStreamSinkChannel.java:257)
	at io.undertow.server.protocol.framed.AbstractFramedStreamSinkChannel.shutdownWrites(AbstractFramedStreamSinkChannel.java:240)
	at io.undertow.websockets.core.WebSockets.flushChannelAsync(WebSockets.java:928)
	at io.undertow.websockets.core.WebSockets.sendInternal(WebSockets.java:912)
	at io.undertow.websockets.core.WebSockets.sendInternal(WebSockets.java:900)
	at io.undertow.websockets.core.WebSockets.sendClose(WebSockets.java:856)
	at io.undertow.websockets.core.WebSockets.sendClose(WebSockets.java:842)
	at io.undertow.websockets.core.WebSockets.sendClose(WebSockets.java:770)
	at io.undertow.websockets.jsr.UndertowSession.closeInternal(UndertowSession.java:211)
	at io.undertow.websockets.jsr.UndertowSession$3$1.run(UndertowSession.java:401)
	at org.xnio.nio.WorkerThread.safeRun(WorkerThread.java:582)
	at org.xnio.nio.WorkerThread.run(WorkerThread.java:466)

11:08:45,405 INFO  (main) [io.undertow.websockets.jsr] <ServerWebSocketContainer.java:747> UT026005: Adding programmatic server endpoint class io.undertow.websockets.jsr.test.JsrWebSocketServer07Test$11TestEndPoint for path /
11:08:45,413 DEBUG (XNIO-1 task-3) [io.undertow.request.security] <SecurityContextImpl.java:86> Attempting to authenticate /, authentication required: false
11:08:45,413 DEBUG (XNIO-1 task-3) [io.undertow.request.security] <SecurityContextImpl.java:247> Authentication outcome was NOT_ATTEMPTED with method io.undertow.security.impl.CachedAuthenticatedSessionMechanism@5621f0ae for /
11:08:45,413 DEBUG (XNIO-1 task-3) [io.undertow.request.security] <SecurityContextImpl.java:110> Authentication result was ATTEMPTED for /


```

**Comment by undertow-pull-request** [comment]
**Linux** [Build 2876](https://ci.wildfly.org/viewLog.html?buildId=134584&buildTypeId=UT_MASTER_LINUX) outcome was **FAILURE** using a merge of c69c7ff6677fb096b05008ebbf0c7363ed6b388f
Summary: Execution timeout (new); tests passed: 3368, ignored: 631 Build time: 01:20:07

**Comment by cakofony** [comment]
The windows failure doesn't appear related to this change, I don't have access to details on the linux build failure.

File: core/src/main/java/io/undertow/io/UndertowOutputStream.java [hunk 1]
```diff
@@ -80,11 +81,9 @@ public void resetBuffer() {
         if(anyAreSet(state, FLAG_WRITE_STARTED)) {
             throw UndertowMessages.MESSAGES.cannotResetBuffer();
         }
-        if(pooledBuffer != null) {
-            pooledBuffer.close();
-            pooledBuffer = null;
```
Comment (cakofony) [inline core/src/main/java/io/undertow/io/UndertowOutputStream.java]: This only nulls the `PooledByteBuffer`, however access using the `buffer()` utility method checks that `buffer` is not null, which this method does not clear, resulting in UndertowOutputStream using a buffer that has potentially been acquired by something else!

I'd really appreciate review when you have a moment @stuartwdouglas, thanks!

**Pull request merged**